source/_posts/analyzing-javascript-with-babel-traverse.md

```diff
@@ -0,0 +1,182 @@
+- It's being assigned to a variable (indicated by the parent being a [`VariableDeclarator`](https://github.com/babel/babel/blob/master/packages/babel-types/README.md#variabledeclarator).
+
+#### Find all the references to styles
+as in `myStyleSheet.someStyle`
+
+`babel-traverse` actually makes this super easy for us, because it **already tracks all variable references**. We can just hang on to the [`Binding`](https://github.com/babel/babel/blob/master/packages/babel-traverse/src/scope/binding.js#L14) object that it uses to collect references, and then iterate through them after traversal.
+
+```js
+const styleSheets = []
+traverse(ast, {
+  CallExpression(path) {
+    if (isStyleSheetCreate(path)) {
+      const members = path.node.arguments[0].properties.filter(
+        property => property.type === 'ObjectProperty' // Not gonna try to figure out spreads
+      )
+      const styleNames = members.map(member => member.key.name)
+      styleSheets.push({
+        id: path.parent.id,
+        styleNames,
+        // hang on to the binding object for "myStyleSheet"
+        binding: path.scope.getBinding(path.parent.id.name)
+      })
+    }
+  },
+})
+```
+
+Now we iterate through our stylesheets, and take a look at the references.
+```js
+styleSheets.forEach(({id, styleNames, binding}) => {
+  const referencedNames = binding.referencePaths
+    .filter(ref => ref.parent.type === 'MemberExpression' && !ref.parent.computed)
+    .map(ref => ref.parent.property.name)
+})
+```
+
+In the example reference `<View style={styles.header} />`, the `referencePath` just points to the `styles` identifier, which is the reference to our stylesheet binding. We want to get the string `"header"`, so we
+
+- determine that we're looking at a [`MemberExpression`](https://github.com/babel/babel/blob/master/packages/babel-types/README.md#memberexpression) (e.g. `styles.something`)
+- make sure it's not computed (we can't do much with `styles[someVariable]`)
+- get the property name (e.g. `"header"`)
+
+#### Then use the list of references to determine which styles aren't being used at all.
+
+```js
+styleSheets.forEach(({id, styleNames, binding}) => {
+  const referencedNames = binding.referencePaths
+    .filter(ref => ref.parent.type === 'MemberExpression' && !ref.parent.computed)
+    .map(ref => ref.parent.property.name)
+  const unused = styleNames.filter(name => referencedNames.indexOf(name) === -1) // 🎉
+})
+```
+
+Just for kicks, we can also get a list of styles that are referenced, but never defined!
+
+
+```js
+styleSheets.forEach(({id, styleNames, binding}) => {
+  const referencedNames = binding.referencePaths
+    .filter(ref => ref.parent.type === 'MemberExpression' && !ref.parent.computed)
+    .map(ref => ref.parent.property.name)
+  const unused = styleNames.filter(name => referencedNames.indexOf(name) === -1)
+  // too easy
+  const missing = referencedNames.filter(name => styleNames.indexOf(name) === -1)
+})
+```
+
+### And we're done!
```

Comment:
May be out of scope for the post, but could you integrate this into ESLint's autofixer? `babel-minify`?